### PR TITLE
Bluetooth: audio: ascs: Retry ASE state notifications on error

### LINKED ISSUE #64574: LE Audio: ASCS notify may run out of buffers
**Describe the bug**
`ase_status_changed` is called for each state change for each ASE. In the case that a client modifies multiple ASEs in a single control point operation, ASCS will attempt to send multiple ASE change notifications in a short burst, requiring a buffer `per` ASE. 

When `bt_gatt_notify` is called it eventually reaches `bt_att_chan_create_pdu` which calls `tx_meta_data_alloc` with `timeout = K_FOREVER`, but since the ASCS state changes are done in the system workqueue, we hit the `if` in `tx_meta_data_alloc`:

```c
static struct bt_att_tx_meta_data *tx_meta_data_alloc(k_timeout_t timeout)
{
	/* The meta data always get freed in the system workqueue,
	 * so if we're in the same workqueue but there are no immediate
	 * contexts available, there's no chance we'll get one by waiting.
	 */
	if (k_current_get() == &k_sys_work_q.thread) {
		return k_fifo_get(&free_att_tx_meta_data, K_NO_WAIT);
	}

	return k_fifo_get(&free_att_tx_meta_data, timeout);
}
```

Which uses `K_NO_WAIT`. This means that `bt_gatt_notify` is not blocking and may fail, and we then hit the following assert if we run out of buffers:
```c
			err = bt_gatt_notify(conn, ase->attr, ase_buf.data, ntf_size);
			__ASSERT_NO_MSG(err == 0);
```

**To Reproduce**
1. Flash the `unicast_audio_server` and `unicast_audio_client` samples
2. Run the samples
3. See the following:
```
QoS: stream 0x2000236c qos 0x200042e8
QoS: interval 10000 framing 0x00 phy 0x02 sdu 40 rtn 2 latency 10 pd 40000
[00:00:35.133,361] <dbg> bt_ascs: ascs_cp_rsp_add: id 0x02 code Success (0x00) reason None (0x00)
[00:00:35.133,575] <dbg> bt_ascs: ase_status_changed: ase 0x20001ed0 id 0x03 codec-configured -> qos-configured
[00:00:35.133,666] <dbg> bt_ascs: ascs_ep_get_status: ep 0x20001ed4 id 0x03 state qos-configured
[00:00:35.133,697] <dbg> bt_ascs: ascs_ep_get_status_qos: dir source codec id 0x06 interval 10000 framing 0x00 phy 0x02 rtn 2 latency 10 pd 40000
[00:00:35.133,911] <dbg> bt_ascs: ase_status_changed: ase 0x20001d50 id 0x01 codec-configured -> qos-configured
[00:00:35.133,941] <dbg> bt_ascs: ascs_ep_get_status: ep 0x20001d54 id 0x01 state qos-configured
[00:00:35.133,972] <dbg> bt_ascs: ascs_ep_get_status_qos: dir sink codec id 0x06 interval 10000 framing 0x00 phy 0x02 rtn 2 latency 10 pd 40000
[00:00:35.134,185] <dbg> bt_ascs: ase_status_changed: ase 0x20001e10 id 0x02 codec-configured -> qos-configured
[00:00:35.134,246] <dbg> bt_ascs: ascs_ep_get_status: ep 0x20001e14 id 0x02 state qos-configured
[00:00:35.134,277] <dbg> bt_ascs: ascs_ep_get_status_qos: dir sink codec id 0x06 interval 10000 framing 0x00 phy 0x02 rtn 2 latency 10 pd 40000
[00:00:35.134,338] <wrn> bt_att: bt_att_chan_create_pdu: Unable to allocate ATT TX meta
[00:00:35.134,368] <wrn> bt_gatt: gatt_notify: No buffer available to send notification
```

**Expected behavior**
Expect ASCS to work with any number of ASEs with any number of TX buffers. 

**Impact**
Annoyance, since it can be circumvented by ensuring that `CONFIG_BT_L2CAP_TX_BUF_COUNT >= CONFIG_BT_ASCS_ASE_SNK_COUNT + CONFIG_BT_ASCS_ASE_SRC_COUNT`. 

**Logs and console output**
See above

**Environment (please complete the following information):**
 - Commit SHA or Version used: c243991000be2711854bf7d66c91f5bf160c0078

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by MariuszSkamra**
This adds retry mechanism for ASE state notifications if failed due to
insufficient number of buffers to send ATT PDU. The state transition is
retried after connection interval delay.

Fixes: #64574

### CODE REVIEW

File: subsys/bluetooth/audio/ascs.c [hunk 1]
```diff
@@ -387,52 +396,92 @@ static void ase_set_state_releasing(struct bt_ascs_ase *ase)
...
+	/* Notify ASE state */
+	if (ase->conn != NULL) {
+		struct bt_conn_info info;
+
+		err = ase_state_notify(ase);
+		if (err == -ENOMEM) {
+			uint32_t retry_delay_ms;
+
+			/* Revert back to old state */
+			ase->ep.status.state = old_state;
```
Comment (Thalley) [inline subsys/bluetooth/audio/ascs.c:419]: Not specific for notifications, but how do we handle this if a client sends another request while we already have a state pending? Do we auto-reject any additional requests, or do we overwrite the pending state?

File: subsys/bluetooth/audio/ascs.c [hunk 2]
```diff
@@ -387,52 +396,92 @@ static void ase_set_state_releasing(struct bt_ascs_ase *ase)
...
+	ase->ep.status.state = new_state;
+
+	/* Notify ASE state */
+	if (ase->conn != NULL) {
+		struct bt_conn_info info;
+
+		err = ase_state_notify(ase);
+		if (err == -ENOMEM) {
+			uint32_t retry_delay_ms;
+
```
Comment (Thalley) [inline subsys/bluetooth/audio/ascs.c:417]:
```suggestion
		err = ase_state_notify(ase);
		if (err == -ENOMEM) {
			struct bt_conn_info info;
			uint32_t retry_delay_ms;

```

File: tests/bluetooth/audio/ascs/src/main.c [hunk 3]
```diff
@@ -648,3 +648,84 @@ ZTEST_F(ascs_test_suite, test_cis_link_loss_in_enabling_state_client_retries)
...
+	if (IS_ENABLED(CONFIG_BT_ASCS_ASE_SNK)) {
+		ase = fixture->ase_snk.attr;
+		ase_id = fixture->ase_snk.id;
+	} else {
+		ase = fixture->ase_src.attr;
+		ase_id = fixture->ase_src.id;
+	}
+
+	zexpect_not_null(ase);
+	zexpect_true(ase_id != 0x00);
```
Comment (Thalley) [inline tests/bluetooth/audio/ascs/src/main.c:690]: Do we have any preference between this and `zassert_equal(ase_id, 0x00)`?

**Review by Thalley — CHANGES_REQUESTED** [review]
Minor comment, otherwise LGTM. 

Have you also verified that the sample works out of the box again now?

File: tests/bluetooth/audio/ascs/src/main.c (hunk 3 above)
Comment (MariuszSkamra) [reply tests/bluetooth/audio/ascs/src/main.c:690]: I don't think so, but your suggestion would be better here

Comment (MariuszSkamra) [reply tests/bluetooth/audio/ascs/src/main.c:690]: > I don't think so, but zassert_not_equal would be better here

**Comment by MariuszSkamra** [comment]
> Minor comment, otherwise LGTM.
> 
> Have you also verified that the sample works out of the box again now?

I've run the BSIM test for unicast client sample 
`./tests/bsim/bluetooth/audio_samples/unicast_audio_client/tests_scripts/unicast_client.sh`

```
d_01: @00:00:05.774560  Setting stream QoS
d_01: @00:00:05.774560  QoS: waiting for 3 streams
d_00: @00:00:05.824352  QoS: stream 0x81ee700 qos 0xf46fcd14
d_00: @00:00:05.824352  QoS: interval 10000 framing 0x00 phy 0x02 sdu 40 rtn 2 latency 10 pd 40000
d_00: @00:00:05.824352  QoS: stream 0x81ee6a0 qos 0xf46fcd14
d_00: @00:00:05.824352  QoS: interval 10000 framing 0x00 phy 0x02 sdu 40 rtn 2 latency 10 pd 40000
d_00: @00:00:05.824352  QoS: stream 0x81ee6c4 qos 0xf46fcd14
d_00: @00:00:05.824352  QoS: interval 10000 framing 0x00 phy 0x02 sdu 40 rtn 2 latency 10 pd 40000
d_00: @00:00:05.824352  [00:00:05.824,340] <wrn> bt_att: Unable to allocate ATT TX meta
d_00: @00:00:05.824352  [00:00:05.824,340] <wrn> bt_gatt: No buffer available to send notification
d_00: @00:00:05.824352  [00:00:05.824,340] <wrn> bt_ascs: Out of buffers for ase state notification. Will retry in 50ms
d_01: @00:00:05.874081  Audio Stream 0x820e530 QoS set
d_01: @00:00:05.874081  QoS: waiting for 3 streams
d_01: @00:00:05.874584  Audio Stream 0x820e4e0 QoS set
d_01: @00:00:05.874584  QoS: waiting for 3 streams
d_01: @00:00:05.923604  Audio Stream 0x820e508 QoS set
d_01: @00:00:05.923604  Stream QoS Set
d_01: @00:00:05.923604  Enabling streams
```

as one can see, the notification has been received in `05.874081`

**Comment by MariuszSkamra** [comment]
> Minor comment, otherwise LGTM.
> Have you also verified that the sample works out of the box again now?

Tested on real HW as well (`unicast_audio_server` vs `unicast_audio_client`)

**Review by Thalley — APPROVED** [review]

**Review by larsgk — APPROVED** [review]

**Review by fredrikdanebjer — APPROVED** [review]

File: subsys/bluetooth/audio/ascs.c [hunk 4]
```diff
@@ -124,56 +124,51 @@ static void ase_free(struct bt_ascs_ase *ase)
...
 
-	(void)k_work_cancel(&ase->state_transition_work);
+	(void)k_work_cancel_delayable(&ase->disconnect_work);
+	(void)k_work_cancel_delayable(&ase->state_transition_work);
 }
 
-static void ase_status_changed(struct bt_ascs_ase *ase, uint8_t state)
+static int ase_state_notify(struct bt_ascs_ase *ase)
 {
+	const uint8_t att_ntf_header_size = 3; /* opcode (1) + handle (2) */
```
[the remark this answers is not in the document]
Comment (Thalley) [reply subsys/bluetooth/audio/ascs.c:133]: @fredrikdanebjer 
https://github.com/zephyrproject-rtos/zephyr/issues/64175

**Pull request merged**